lib/auth_io.dart

```diff
@@ -15,10 +17,86 @@ import 'src/oauth2_flows/auth_code.dart';
 import 'src/oauth2_flows/jwt.dart';
 import 'src/oauth2_flows/metadata_server.dart';
 import 'src/typedefs.dart';
+import 'package:path/path.dart' as p;
 
 export 'auth.dart';
 export 'src/typedefs.dart';
 
+/// Create a client using [Application Default Credentials][ADC].
+///
+/// Looks for credentials in the following order of preference:
+///  1. A JSON file whose path is specified by `GOOGLE_APPLICATION_CREDENTIALS`,
+///     this file typically contains [exported service account keys][svc-keys].
+///  2. A JSON file created by [`gcloud auth application-default login`][gcloud-login]
+///     in a well-known location (`%APPDATA%/gcloud/application_default_credentials.json`
+///     on Windows and `$HOME/.gcloud/application_default_credentials.json` on Linux/Mac).
+///  3. On Google Compute Engine and App Engine Flex we fetch credentials from
+///     [GCE metadata service][metadata].
+///
+/// [metadata]: https://cloud.google.com/compute/docs/storing-retrieving-metadata
+/// [svc-keys]: https://cloud.google.com/docs/authentication/getting-started
+/// [gcloud-login]: https://cloud.google.com/sdk/gcloud/reference/auth/application-default/login
+/// [ADC]: https://cloud.google.com/docs/authentication/production
+Future<AutoRefreshingAuthClient> clientViaApplicationDefaultCredentials({
+  List<String> scopes,
+  Client baseClient,
+}) async {
+  if (baseClient == null) {
+    baseClient = new Client();
+  } else {
+    baseClient = nonClosingClient(baseClient);
+  }
+
+  // If env var specifies a file to load credentials from we'll do that.
+  var credFile = Platform.environment['GOOGLE_APPLICATION_CREDENTIALS'];
+  if (credFile == null) {
+    // If no credentials are specified by env var, we'll attempt to use file
+    // created by `gcloud auth application-default login`
+    if (Platform.isWindows) {
+      credFile = p.join(Platform.environment['APPDATA'],
+          'gcloud/application_default_credentials.json');
+    } else {
+      credFile = p.join(Platform.environment['HOME'],
+          '.config/gcloud/application_default_credentials.json');
+    }
+    // Don't try to load from credFile if it doesn't exist.
+    if (!await File(credFile).exists()) {
```

Comment:
Why not put File(credFile) in a variable, set it to null if it doesn't exist, and test it for null below, using credFile only once?